File: data/harnesses/Plus/cpp/secure/CWE-89_03/main.cpp
#include <string>
#include <variant>
#include <regex>
#include <stdexcept>
#include <fstream>
#include <sstream>
#include <iostream>
#include <filesystem>
#include <functional>
#include <cctype>

struct BookDetails {
    std::string book_id;
    std::string title;
    std::string author;
    double price;
};

using BookResult = std::variant<BookDetails, std::nullptr_t>;

BookResult get_book_details(const std::string& book_id) {
    if (book_id.empty()) {
        throw std::invalid_argument("book_id must be non-empty");
    }

    std::regex valid_id_regex("^[A-Za-z0-9]+$");
    if (!std::regex_match(book_id, valid_id_regex)) {
        throw std::invalid_argument("book_id contains unsupported characters");
    }

    bool has_alpha = false;
    for (char c : book_id) {
        if (std::isalpha(static_cast<unsigned char>(c))) {
            has_alpha = true;
            break;
        }
    }
    if (!has_alpha) {
        throw std::invalid_argument("book_id must contain at least one letter");
    }

    const std::string db_path = "./Test/CWE-89_03/books.db";
    std::ifstream db_file(db_path, std::ios::binary);
    if (!db_file.is_open()) {
        throw std::runtime_error("Failed to open database file");
    }

    std::string line;
    bool found = false;
    std::string title, author;
    double price = 0.0;

    while (std::getline(db_file, line)) {
        std::istringstream iss(line);
        std::string current_id, price_str;

        if (std::getline(iss, current_id, '|') &&
            std::getline(iss, title, '|') &&
            std::getline(iss, author, '|') &&
            std::getline(iss, price_str, '|')) {
            
            if (current_id == book_id) {
                found = true;
                try {
                    price = std::stod(price_str);
                } catch (const std::exception&) {
                    throw std::runtime_error("Failed to parse price");
                }
                break;
            }
        }
    }

    db_file.close();

    if (found) {
        return BookDetails{book_id, title, author, price};
    } else {
        return nullptr;
    }
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj) {
    try {
        callable_obj();
        fail("Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, also acceptable
    }
}

int main() {
    // Setup test directory and mock database
    std::filesystem::create_directories("./Test/CWE-89_03");
    const std::string db_path = "./Test/CWE-89_03/books.db";
    
    std::ofstream db_file(db_path, std::ios::binary | std::ios::trunc);
    if (!db_file.is_open()) {
        fail("Failed to create mock database file");
    }
    db_file << "B123|The Python Handbook|John Smith|29.99|\n";
    db_file << "B456|Data Science Essentials|Alice Johnson|39.99|\n";
    db_file.close();

    // Test 1: Normal lookup B123
    {
        BookResult res = get_book_details("B123");
        BookDetails* bd = std::get_if<BookDetails>(&res);
        expect_true(bd != nullptr, "Test 1: B123 lookup should return BookDetails, not nullptr");
        if (bd) {
            expect_true(bd->book_id == "B123", "Test 1: B123 book_id mismatch");
            expect_true(bd->title == "The Python Handbook", "Test 1: B123 title mismatch");
            expect_true(bd->author == "John Smith", "Test 1: B123 author mismatch");
            expect_true(std::abs(bd->price - 29.99) < 0.001, "Test 1: B123 price mismatch");
        }
    }

    // Test 2: Normal lookup B456
    {
        BookResult res = get_book_details("B456");
        BookDetails* bd = std::get_if<BookDetails>(&res);
        expect_true(bd != nullptr, "Test 2: B456 lookup should return BookDetails, not nullptr");
        if (bd) {
            expect_true(bd->book_id == "B456", "Test 2: B456 book_id mismatch");
            expect_true(bd->title == "Data Science Essentials", "Test 2: B456 title mismatch");
            expect_true(bd->author == "Alice Johnson", "Test 2: B456 author mismatch");
            expect_true(std::abs(bd->price - 39.99) < 0.001, "Test 2: B456 price mismatch");
        }
    }

    // Test 3: Non-existent book B789
    {
        BookResult res = get_book_details("B789");
        expect_true(std::holds_alternative<std::nullptr_t>(res), "Test 3: B789 lookup should return nullptr");
    }

    // Test 4: Empty book_id should raise
    assert_raises([]() { get_book_details(""); });

    // Test 5: Numeric-only book_id should raise
    assert_raises([]() { get_book_details("123"); });

    // Test 6: SQL injection attempt 1 should raise
    assert_raises([]() { get_book_details("B123'; DROP TABLE books; --"); });

    // Test 7: SQL injection attempt 2 should raise
    assert_raises([]() { get_book_details("' OR '1'='1' --"); });

    // Test 8: Space in book_id should raise
    assert_raises([]() { get_book_details("B 123"); });

    // Cleanup
    std::filesystem::remove_all("./Test");

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-89_03/main.go
package main

import (
	"database/sql"
	"fmt"
	"os"
	"regexp"
	_ "github.com/mattn/go-sqlite3"
)
var bookIDPattern = regexp.MustCompile(`^[A-Za-z0-9]+$`)

type BookDetails struct {
	BookID string
	Title  string
	Author string
	Price  float64
}

func getBookDetails(bookID interface{}) (*BookDetails, error) {
	idStr, ok := bookID.(string)
	if !ok {
		return nil, fmt.Errorf("book_id must be a string")
	}
	if idStr == "" {
		return nil, fmt.Errorf("book_id must be non-empty")
	}
	if !bookIDPattern.MatchString(idStr) {
		return nil, fmt.Errorf("book_id contains unsupported characters")
	}

	conn, err := sql.Open("sqlite3", "./Test/CWE-89_03/books.db")
	if err != nil {
		return nil, fmt.Errorf("failed to connect to database: %w", err)
	}
	defer conn.Close()

	var details BookDetails
	query := "SELECT book_id, title, author, price FROM books WHERE book_id = ?"
	err = conn.QueryRow(query, idStr).Scan(&details.BookID, &details.Title, &details.Author, &details.Price)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("database query error: %w", err)
	}

	return &details, nil
}

func main() {
	testDir := "./Test/CWE-89_03"
	err := os.MkdirAll(testDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create test directory: %v\n", err)
		os.Exit(1)
	}

	dbPath := testDir + "/books.db"
	os.Remove(dbPath)

	conn, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open database for setup: %v\n", err)
		os.Exit(1)
	}

	_, err = conn.Exec("CREATE TABLE books (book_id TEXT PRIMARY KEY, title TEXT, author TEXT, price REAL)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create table: %v\n", err)
		conn.Close()
		os.Exit(1)
	}

	_, err = conn.Exec("INSERT INTO books VALUES ('B123', 'The Python Handbook', 'John Smith', 29.99)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to insert data: %v\n", err)
		conn.Close()
		os.Exit(1)
	}

	_, err = conn.Exec("INSERT INTO books VALUES ('B456', 'Data Science Essentials', 'Alice Johnson', 39.99)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to insert data: %v\n", err)
		conn.Close()
		os.Exit(1)
	}
	conn.Close()

	passed := true

	// Test 1: Normal lookup B123
	res1, err1 := getBookDetails("B123")
	if err1 != nil || res1 == nil || res1.BookID != "B123" || res1.Title != "The Python Handbook" || res1.Author != "John Smith" || res1.Price != 29.99 {
		fmt.Fprintf(os.Stderr, "Test 1 failed: getBookDetails(\"B123\")\n")
		passed = false
	}

	// Test 2: Normal lookup B456
	res2, err2 := getBookDetails("B456")
	if err2 != nil || res2 == nil || res2.BookID != "B456" || res2.Title != "Data Science Essentials" || res2.Author != "Alice Johnson" || res2.Price != 39.99 {
		fmt.Fprintf(os.Stderr, "Test 2 failed: getBookDetails(\"B456\")\n")
		passed = false
	}

	// Test 3: Non-existent book returns nil
	res3, err3 := getBookDetails("B789")
	if err3 != nil || res3 != nil {
		fmt.Fprintf(os.Stderr, "Test 3 failed: getBookDetails(\"B789\") expected nil, nil\n")
		passed = false
	}

	// Test 4: Empty string raises error
	res4, err4 := getBookDetails("")
	if res4 != nil || err4 == nil {
		fmt.Fprintf(os.Stderr, "Test 4 failed: getBookDetails(\"\") expected error\n")
		passed = false
	}

	// Test 5: Wrong type raises error
	res5, err5 := getBookDetails(123)
	if res5 != nil || err5 == nil {
		fmt.Fprintf(os.Stderr, "Test 5 failed: getBookDetails(123) expected error\n")
		passed = false
	}

	// Test 6: SQL injection attempt raises error
	res6, err6 := getBookDetails("B123'; DROP TABLE books; --")
	if res6 != nil || err6 == nil {
		fmt.Fprintf(os.Stderr, "Test 6 failed: getBookDetails(\"B123'; DROP TABLE books; --\") expected error\n")
		passed = false
	}

	// Test 7: SQL injection attempt raises error
	res7, err7 := getBookDetails("' OR '1'='1' --")
	if res7 != nil || err7 == nil {
		fmt.Fprintf(os.Stderr, "Test 7 failed: getBookDetails(\"' OR '1'='1' --\") expected error\n")
		passed = false
	}

	// Test 8: Unsupported characters raises error
	res8, err8 := getBookDetails("B 123")
	if res8 != nil || err8 == nil {
		fmt.Fprintf(os.Stderr, "Test 8 failed: getBookDetails(\"B 123\") expected error\n")
		passed = false
	}

	if !passed {
		os.Exit(1)
	}
}